### PR TITLE
add `emit_excinfo_store_with_md` for all `excinfo`-related `store` instructions

### LINKED ISSUE #9550: LLVM IR generation bug when adding metadata for `excinfo`-related store instruction
I noticed in a certain case, numba will emit LLVM IR like this:
```llvm
B10.if:                                           ; preds = %B10.if.loopexit, %B10.if.loopexit10
  %excinfo.2.0 = phi { i8*, i32, i8*, i8*, i32 }* [ @.const.picklebuf.23377073394240, %B10.if.loopexit10 ], [ @.const.picklebuf.23377075691904, %B10.if.loopexit ]
  %2 = bitcast i32* %.9.i to i8*
  %3 = bitcast i32* %.7.i to i8*
  call void @llvm.lifetime.end.p0i8(i64 4, i8* nonnull %3)
  call void @llvm.lifetime.end.p0i8(i64 4, i8* nonnull %2)
  store { i8*, i32, i8*, i8*, i32 }* %excinfo.2.0, { i8*, i32, i8*, i8*, i32 }** %excinfo, align 8
  br label %common.ret
```
I think this branch is used for storing `excinfo` and return. But the `store` instr doesn't have the metadata of `numba_exception_output`, which makes the `ref_prune` pass in llvmlite can't remove some ref pairs. 
The source code in llvmlite:
```c++
    bool isRaising(const BasicBlock *bb) {
        for (auto &instruction : *bb) {
            if (instruction.getOpcode() == Instruction::Store) {
                auto name = dyn_cast<StoreInst>(&instruction)
                                ->getPointerOperand()
                                ->getName();
                if (name.equals("excinfo") &&
                    instruction.hasMetadata("numba_exception_output")) {
                    return true;
                }
            }
        }
        return false;
    }
```
Willing to provide a python test case later.

### PR DESCRIPTION
**Pull request by dlee992**
fixes #9550 

This bug is introduced between numba upgraded to llvm14 (https://github.com/numba/numba/pull/8535) and added the dynamic exception info feature (https://github.com/numba/numba/pull/8134).

### CODE REVIEW

**Comment by dlee992** [comment]
BTW, can I be upgraded to a collaborator? or can I be given the permission about marking issues and PRs? I can only use this permission to mark the issues and PRs opened by myself.

Ha, just asking...

**Comment by dlee992** [comment]
Oh, if possible, I think this fix should be put in release ASAP, perhaps 0.60?
For my use-cases (heavily using various numba containers and structref), my performance could be hurt by 2~3x times, since un-removed refcount pairs.

**Comment by esc** [comment]
> Oh, if possible, I think this fix should be put in release ASAP, perhaps 0.60? For my use-cases (heavily using various numba containers and structref), my performance could be hurt by 2~3x times, since un-removed refcount pairs.

This release is closed now, unfortunately. I've restarted CI and marked this as needing a review.

**Comment by esc** [comment]
> BTW, can I be upgraded to a collaborator? or can I be given the permission about marking issues and PRs? I can only use this permission to mark the issues and PRs opened by myself.
> 
> Ha, just asking..

Would suggest to join a dev meeting or office hour and ask about this.

File: numba/core/callconv.py [hunk 1]
```diff
@@ -394,6 +394,11 @@ def build_excinfo_struct(self, exc, exc_args, loc, func_name):
         exc = (exc, exc_args, locinfo)
         return exc
 
+    def emit_excinfo_store_with_md(self, builder, value, excinfo_ptr):
+        store = builder.store(value, excinfo_ptr)
+        md = builder.module.add_metadata([ir.IntType(1)(1)])
+        store.set_metadata("numba_exception_output", md)
+
```
[the remark this answers is not in the document]
Comment (dlee992) [reply numba/core/callconv.py:401]: Yeah, sure, but why we need to keep `store` out?

File: numba/core/callconv.py [hunk 2]
```diff
@@ -762,12 +765,12 @@ def unset_try_status(self, builder):
         # will run normally.
         excinfoptr = self._get_excinfo_argument(builder.function)
         null = cgutils.get_null_value(excinfoptr.type.pointee)
-        builder.store(null, excinfoptr)
+        self.emit_excinfo_store_with_md(builder, null, excinfoptr)
```
[the remark this answers is not in the document]
Comment (dlee992) [reply numba/core/callconv.py:768]: Good question! But I didn't know the answer. I just did this conservatively.

**Comment by dlee992** [comment]
I think if the jitted code is related to structref or container types, then the code logic can touch the dynamic exception info generation, then we can find a simple test case for it. Need to make up one.

**Comment by dlee992** [comment]
reopen...

**Pull request closed without merging**